File: dashboard/src/api/queries/goals.ts
import { api } from "../client"
import type { Goal } from "../../types"

export async function fetchGoals(): Promise<Goal[]> {
  const { data } = await api.get("/api/goals")
  return data
}

export async function createGoal(goal: { title: string; time_horizon?: string; category?: string; description?: string }): Promise<void> {
  await api.post("/api/goals", goal)
}

export async function updateGoal(id: number, updates: Partial<Goal>): Promise<void> {
  await api.patch(`/api/goals/${id}`, updates)
}

export async function completeGoal(id: number): Promise<void> {
  await api.post(`/api/goals/${id}/complete`)
}

export async function deleteGoal(id: number): Promise<void> {
  await api.delete(`/api/goals/${id}`)
}

export async function addSubtask(goalId: number, title: string): Promise<void> {
  await api.post(`/api/goals/${goalId}/subtasks`, { title })
}

export async function toggleSubtask(subtaskId: number): Promise<void> {
  await api.post(`/api/goals/subtasks/${subtaskId}/toggle`)
}

export async function deleteSubtask(subtaskId: number): Promise<void> {
  await api.delete(`/api/goals/subtasks/${subtaskId}`)
}

File: lora_api/routers/goals.py
from fastapi import APIRouter, Depends
from lora_api.auth import get_current_user
from lora_api.database import get_pool
from lora_api.serializers import clean_dict

router = APIRouter(prefix="/api", tags=["goals"])


@router.get("/goals")
async def list_goals(user=Depends(get_current_user)):
    import db.queries.goals as q
    pool = await get_pool()
    rows = await q.get_all_goals(pool)
    return [clean_dict(dict(r)) for r in rows]


@router.post("/goals")
async def create_goal(data: dict, user=Depends(get_current_user)):
    import db.queries.goals as q
    pool = await get_pool()
    gid = await q.add_goal(
        pool, title=data["title"],
        description=data.get("description"),
        category=data.get("category", "general"),
        time_horizon=data.get("time_horizon", "month"),
    )
    return {"id": gid, "status": "created"}


@router.patch("/goals/{goal_id}")
async def update_goal(goal_id: int, data: dict, user=Depends(get_current_user)):
    import db.queries.goals as q
    pool = await get_pool()
    await q.update_goal(pool, goal_id, **data)
    return {"status": "updated"}


@router.post("/goals/{goal_id}/complete")
async def complete_goal(goal_id: int, user=Depends(get_current_user)):
    import db.queries.goals as q
    pool = await get_pool()
    await q.complete_goal(pool, goal_id)
    return {"status": "completed"}


@router.delete("/goals/{goal_id}")
async def delete_goal(goal_id: int, user=Depends(get_current_user)):
    import db.queries.goals as q
    pool = await get_pool()
    await q.delete_goal(pool, goal_id)
    return {"status": "deleted"}


@router.post("/goals/{goal_id}/subtasks")
async def add_subtask(goal_id: int, data: dict, user=Depends(get_current_user)):
    import db.queries.goals as q
    pool = await get_pool()
    sid = await q.add_subtask(pool, goal_id, data["title"])
    return {"id": sid, "status": "created"}


@router.post("/goals/subtasks/{subtask_id}/toggle")
async def toggle_subtask(subtask_id: int, user=Depends(get_current_user)):
    import db.queries.goals as q
    pool = await get_pool()
    await q.complete_subtask(pool, subtask_id)
    return {"status": "toggled"}


@router.delete("/goals/subtasks/{subtask_id}")
async def delete_subtask(subtask_id: int, user=Depends(get_current_user)):
    import db.queries.goals as q
    pool = await get_pool()
    await q.delete_subtask(pool, subtask_id)
    return {"status": "deleted"}
